### PR TITLE
WT-12435 Add statistic that tracks how many API calls are in flight

### PR DESCRIPTION
**Pull request by agorrod**
This PR introduces two new statistics, one that tracks all threads inside the library, the other that tracks threads currently doing cursor operations in the library.

A wtperf test with non-json logging shows the following output for these stats:

```
Feb 16 03:53:52 16 WT_TEST connection: number of threads currently inside the library
Feb 16 03:53:52 6 WT_TEST connection: number of threads currently inside the library doing cursor operations
```

It bears some investigation to understand how the broader stats are 16 - it's possible that internal callers are artificially bumping that number?

### CODE REVIEW

**Comment by agorrod** [comment]
For the record, these statistics are aimed at giving us more confidence about how much time is spent doing work inside WiredTiger. The "cursor" version is meant as a metric for "real" work.

The broader threads inside the library will include things like checkpoint and global set_timestamp calls.

Screenshot 2024-02-26 at 11 07 18 am

I need to rework the name of the stats so they are shorter and more palatable, but the above graph seems useful.

The markers A-B show a period of time where the workload throughput drops, it's useful to see that during that same period there isn't more time being spent inside WiredTiger. Which is a strong indicator that the slowdown wasn't due to latencies increasing inside WiredTiger.

I've also included in the graph the MongoDB ticket metrics - it's worth noting that those are present on the primary, but not the secondaries, since oplog batch application doesn't acquire a ticket.

File: src/cursor/cur_file.c
Comment (agorrod) [inline src/cursor/cur_file.c:125]: The pattern here is pretty ugly, in terms of using a tracked variable and repeating that a bunch of times in the file.

I started trying to use an API macro, but gave up due to nuances and complexity in the tracking macros that I didn't want to make worse. Then I realized that sometimes the API tracking macros go to error directly, so this tracking could be skipped, and thus the local variable to know what has happened.
I thought about adding a field to the WT_CURSOR structure, and passing that to the track_cursor API to avoid the local variable, but I got worried that it would be hard to ensure it is reset in all cases..

I'm open to better ideas.

File: src/include/api.h
Comment (agorrod) [inline src/include/api.h:97]: Reviewing this deserves some care: The caller says "No code before this line", I think this violates that. I think it's OK since the code never generates an error return, so the error setup should always complete, but it's worth taking some time to think through

File: src/include/api_inline.h
Comment (agorrod) [inline src/include/api_inline.h:23]: The check against session->id == 0 is sprinkled around this change. It turns out that the default session is used for a bunch more things than it should be. I'd like to fix that and remove the checks, but not within the purview of this change - so I've just excluded it.

File: src/include/api.h
Comment (korteland) [inline src/include/api.h:68]: Is there a macro for this? e.g. `WT_SESSION_ID_INTERNAL`?

File: src/include/api.h
Comment (korteland) [reply src/include/api.h:97]: I agree this is infallible. Whether being infallible is enough to allow an exception to the "no code before this line" comment, I don't know.

Comment (korteland) [inline src/include/api.h:97]: What's the deal with checking `api_call_counter == 1`? I don't understand what this does or the interaction with the new counters.

Comment (korteland) [reply src/include/api.h:97]: Oh, I think I get it now - is it for handling reentrancy?

File: src/cursor/cur_file.c
Comment (korteland) [reply src/cursor/cur_file.c:125]: I think it's worth questioning the premise of not tracking calls that end up failing the setup macro. The stats count threads in API calls, and I don't see why a failed API call setup shouldn't count towards that. 

Theoretically (but hopefully not in practice) the setup macro could take a long time, or get stuck in an infinite loop. And I'd definitely want to see how many threads were in API calls if that happened.

My head isn't as deep into this space as yours, but it appears that would let us drop the `tracked` variables and the explicit `__wt_api_track_cursor_(start|end)` (instead bundling them into the macros).

File: src/include/api_inline.h
Comment (korteland) [inline src/include/api_inline.h:16]: Minor point on naming - begin/end and start/finish feel like more natural "name pairs" than start/end.

File: src/include/api.h
Comment (agorrod) [reply src/include/api.h:97]: Yes - that's right. It pre-existed this change and avoids double counting on re-entrance (which is common)

File: src/cursor/cur_file.c
Comment (agorrod) [reply src/cursor/cur_file.c:125]: I had started there (by nesting this inside the other API tracking macros), but it became too complex.

It's hard to move this outside of the API entry macros - those are supposed to be the first and last things an API does (I'm sure there is at least one exception, but generally). The functionality here also relies on api call counters to track recursion that is tracked via the macros.

It isn't common for a macro to hang before returning an error. For example, I noticed this because some of our Python tests deliberately try to do illegal operations on prepared transactions - failures of which are flagged via the macro.

File: src/include/api.h
Comment (agorrod) [reply src/include/api.h:97]: Thanks - that's enough

**Comment by agorrod** [comment]
@korteland  - thanks for your patience here, and sorry for the long gap between reviews, but I've taken your (great) feedback and made some updates. Would you mind taking another look?

File: src/cursor/cur_file.c
Comment (korteland) [reply src/cursor/cur_file.c:125]: Thanks - I would like to take you up on the Zoom meeting, I think there's something I'm missing about the problems that having the call in the macro causes.

The thing I don't get here is: we have a set of macros that are called at the beginning and end of each interesting function. Even if they jump directly to `err`, all uses call the right `API_END` macro. So I'd have thought that putting the `__wt_api_track_cursor_X` calls towards the beginning of the ~5 relevant `CURSOR_API*` macros, before any failures are possible, would "naturally" result in the correct counting behaviour.

**Review by korteland — APPROVED** [review]
Thanks! This LGTM modulo this thread (https://github.com/wiredtiger/wiredtiger/pull/10231#discussion_r1522520622). And I don't dislike the `bool tracked` stuff enough to hold the PR up :-)

**Comment by korteland** [comment]
Writing down what we discussed online - I now understand that the API macro that has to be called first is the "inner" one, e.g. `TXN_API_CALL_NOCONF`. So now I understand the thinking behind the `bool tracked` approach. These macros need a cleanup so the plan was to add a FIXME for the API macro cleanup ticket.

**Comment by agorrod** [comment]
@ddanderson would you mind reviewing this please? I chose you because you've most recently made changes to api.h - feel free to delegate to someone else if you'd prefer

File: src/include/api.h
Comment (ddanderson) [inline src/include/api.h:68]: Maybe use `!WT_SESSION_IS_DEFAULT(s)`

File: src/include/api.h
Comment (ddanderson) [inline src/include/api.h:77]: `!WT_SESSION_IS_DEFAULT(s)`

File: src/include/api.h
Comment (ddanderson) [inline src/include/api.h:97]: `!WT_SESSION_IS_DEFAULT(s)`.
Maybe swap the order so the check for `(s)->api_call_counter == 1` is first.  (It's more likely to be false in cases we care about in performance, when we're not using the default session).

File: src/session/session_api.c
Comment (ddanderson) [inline src/session/session_api.c:318]: Could do something like:
```
     WT_CONNECTION_IMPL *track_conn;
     bool internal;
....
     internal = F_ISSET(session, WT_SESSION_INTERNAL);
     if (!WT_SESSION_IS_DEFAULT(session) && session->api_call_counter == 1)
        track_conn = S2C(session);
     else
        track_conn = NULL;

      WT_TRET(__wt_session_close_internal(session));
      session = NULL;
      if (track_conn) {
         if (internal)
              wt_atomic_add(...);
         else
              wt_atomic_add(...);
   }
```

File: src/cursor/cur_file.c
Comment (ddanderson) [inline src/cursor/cur_file.c:162]: What would you think of creating three macros:
```
  #define  WT_DECL_CUR_TRACK    bool _tracked = false;
  #define WT_CUR_TRACK_BEGIN(session)  __wt_api_track_cursor_begin(session); _tracked = true
  #define WT_CUR_TRACK_END(session) if (_tracked) __wt_api_track_cursor_end(session)
```
It might appear a bit cleaner, and explanatory comments would be in one place.

Since we have a larger set of these BEGIN/END matching type macros, we could create a script in dist that checks to make sure that, when used, they match up and are ordered as expected.  (Another ticket).

File: src/include/api.h
Comment (ddanderson) [inline src/include/api.h:105]: ditto WT_SESSION_IS_DEFAULT and consider swapping order

**Review by ddanderson — APPROVED** [review]
@agorrod, this is an exciting change, it will be very useful.
I made some suggestions to consider to potentially clean up a couple things.  Take them if you want, no need for re-review.

File: src/session/session_api.c
Comment (agorrod) [reply src/session/session_api.c:318]: Thanks. I have been trying to figure out how accurate this should be. Obviously it would be best if it was entirely accurate, but I also don't think we need to worry too much about cases like this.

From my point of view the internal tracking was really a bit of free information - since I initially didn't distinguish and was surprised that concurrent activity was sometimes higher than workload activity.

The likelihood that a meaningful amount of time is spent inside WiredTiger in session close seems low enough that I'm not concerned about getting it right.

I left this big comment here primarily so that a subsequent API entry/exit macro tracking change would think about getting this right (if possible).

Let me know if you think it's worth getting this right - I'll be happy to take your suggested change (thanks for taking the time).

File: src/cursor/cur_file.c
Comment (agorrod) [reply src/cursor/cur_file.c:162]: It is interesting. I'll try it out.

I moved away from the tracking macro approach since I wanted the functionality to be in a function so I could add diagnostic checks that were accessible in a debugger. Hopefully that isn't necessary after this change is merged.

**Review by agorrod — COMMENTED** [review]
Thanks for your review @ddanderson - I'm a bit slow in responding, but I appreciate it.

I've responded to your bigger questions, but it'll be at least another day before this is ready for another review. I'll let you know when it gets to that.

File: src/session/session_api.c
Comment (ddanderson) [reply src/session/session_api.c:318]: I don't think it's very important, as you say, the close call should be pretty fast.  "Fixing" it does make the code more complex, so I'm heppy with the original PR.

**Comment by agorrod** [comment]
@korteland - would you mind giving this another look over? I've addressed the review comments and would appreciate a final stamp of approval (or RFC) before merging.

**Review by korteland — APPROVED** [review]
LGTM! One question, no need for a re-review.

File: src/include/api_inline.h
Comment (agorrod) [reply src/include/api_inline.h:24]: It is `s_all`'ed.

**Pull request merged**
